File: src/main/java/sx/event/ThreadSafeEventDispatcher.java
package sx.event;

import java.util.List;

/**
 * Threadsafe multicast event dispatcher.
 *
 * @author masc
 */
public class ThreadSafeEventDispatcher<T extends EventListener> extends RegularEventDispatcher<T> {
    @Override
    public synchronized void add(T listener) {
        super.add(listener);
    }

    @Override
    public synchronized void remove(T listener) {
        super.remove(listener);
    }

    @Override
    protected synchronized List<ListenerReference> getListeners() {
        return super.getListeners();
    }

    @Override
    protected synchronized void remove(List<ListenerReference> listeners) {
        super.remove(listeners);
    }
}

File: libs/sx-common/src/main/java/sx/event/ThreadSafeEventDispatcher.kt
package sx.event

import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * Threadsafe multicast event dispatcher.

 * @author masc
 */
class ThreadSafeEventDispatcher<T : EventListener> : RegularEventDispatcher<T>() {
    private val lock = ReentrantLock()

    override fun add(listener: T) {
        this.lock.withLock {
            super.add(listener)
        }
    }

    override fun remove(listener: T) {
        this.lock.withLock {
            super.remove(listener)
        }
    }

    override val listeners: List<ListenerReference>
    get() {
        this.lock.withLock {
            return super.listeners
        }
    }

    override fun remove(listeners: List<ListenerReference>) {
        this.lock.withLock {
            super.remove(listeners)
        }
    }
}
